contracts/external/callees/Trade0x.sol

```diff
@@ -0,0 +1,167 @@
+        weth = ERC20Interface(_weth);
+        weth.safeApprove(_assetProxy, uint256(-1));
+        weth.safeApprove(_staking, uint256(-1));
+        controller = _controller;
+    }
+
+    event Trade(
+        address indexed taker,
+        address indexed maker,
+        address takerAsset,
+        address makerAsset,
+        uint256 takerAmount,
+        uint256 makerAmount
+    );
+
+    /**
+     * @dev fill 0x order
+     * @param _sender the original sender who wants to trade on 0x
+     * @param _data abi-encoded order, fillamount, signature and feePayer. fee payer is the address we pull weth from.
+     */
+
+    function callFunction(address payable _sender, bytes memory _data) external override {
+        require(msg.sender == controller, "sender not controller");
+        _directlyTrade(_sender, _data);
+        // _indirectlyTrade(_sender, _data);
+    }
+
+    function _indirectlyTrade(address payable _sender, bytes memory _data) internal {
+        (IZeroXExchange.Transaction memory transaction, bytes memory signature) = abi.decode(
+            _data,
+            (IZeroXExchange.Transaction, bytes)
+        );
+
+        exchange.executeTransaction(transaction, signature);
+    }
+
+    function _directlyTrade(address payable _sender, bytes memory _data) internal {
+        (
+            IZeroXExchange.Order[] memory order,
+            uint256[] memory takerAssetFillAmount,
+            bytes[] memory signature,
+            address feePayer
+        ) = abi.decode(_data, (IZeroXExchange.Order[], uint256[], bytes[], address));
+
+        for (uint256 i = 0; i < order.length; i++) {
+            address takerAsset = decodeERC20Asset(order[i].takerAssetData);
+            // pull token from user
+            ERC20Interface(takerAsset).safeTransferFrom(_sender, address(this), takerAssetFillAmount[i]);
+
+            // approve the 0x ERC20 Proxy to move fund
+            ERC20Interface(takerAsset).safeIncreaseAllowance(assetProxy, takerAssetFillAmount[i]);
+        }
+
+        // pull weth (to pay 0x) from feePayer address
+        uint256 protocolFee = tx.gasprice * PORTOCAL_FEE_BASE;
+        weth.safeTransferFrom(feePayer, address(this), protocolFee);
+
+        IZeroXExchange.FillResults[] memory result = exchange.batchFillOrders(order, takerAssetFillAmount, signature);
+
+        for (uint256 i = 0; i < order.length; i++) {
+            // transfer swapped token to sender
+            address makerAsset = decodeERC20Asset(order[i].makerAssetData);
+            uint256 balance = ERC20Interface(makerAsset).balanceOf(address(this));
+            ERC20Interface(makerAsset).safeTransfer(_sender, balance);
+
+            // transfer the taker asset back to the user if the order wasn't fully filled
+            address takerAsset = decodeERC20Asset(order[i].takerAssetData);
+            balance = ERC20Interface(takerAsset).balanceOf(address(this));
+            ERC20Interface(takerAsset).safeTransfer(_sender, balance);
```

Comment:
we should probably check if( balance > 0) here? cause users may fill multiple orders with same makerAsset. Otherwise the safeTransfer will fail if we call with amount = 0 (not sure exactly how it's gonna behave)

Reply:
okay, let me test how it behaves on amount = 0.